File: app4web_2022/app/src/main/java/site/app4web/app4web/Lib/MaterialBadgeTextView.java
// A mildly-modified version of the MaterialBadgeTextView library from:
// https://github.com/matrixxun/MaterialBadgeTextView

package site.app4web.app4web.Lib;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.RadialGradient;
import android.graphics.RectF;
import android.graphics.Shader;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.OvalShape;
import android.os.Build;
//import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import site.app4web.app4web.Helper.JasonHelper;

import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.view.ViewCompat;

/**
 * Created by matrixxun on 2016/8/30.
 */
public class MaterialBadgeTextView extends AppCompatTextView {

    private static final int DEFAULT_FILL_TYPE = 0;

    private int backgroundColor;

    private static final float SHADOW_RADIUS = 3.5f;
    private static final int FILL_SHADOW_COLOR = 0x55000000;
    private static final int KEY_SHADOW_COLOR = 0x55000000;

    private static final float X_OFFSET = 0f;
    private static final float Y_OFFSET = 1.75f;


    private int mShadowRadius;
    private int shadowYOffset;
    private int shadowXOffset;

    private int diffWH;

    private boolean isHighLightMode;

    public MaterialBadgeTextView(final Context context) {
        this(context, null);
    }

    public MaterialBadgeTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MaterialBadgeTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        setGravity(Gravity.CENTER);
        float density = getContext().getResources().getDisplayMetrics().density;
        mShadowRadius = (int) (density * SHADOW_RADIUS);
        shadowYOffset = (int) (density * Y_OFFSET);
        shadowXOffset = (int) (density * X_OFFSET);
        int basePadding = (mShadowRadius * 2);
        float textHeight = getTextSize();
        float textWidth = textHeight / 4;
        diffWH = (int) (Math.abs(textHeight - textWidth) / 2);
        int horizontalPadding = basePadding + diffWH;
        setPadding(horizontalPadding, basePadding, horizontalPadding, basePadding);
        backgroundColor = JasonHelper.parse_color("#ff0000");
        int borderColor = Color.TRANSPARENT;
        float borderWidth = 0;
        float borderAlpha = 1;
        int ctType = DEFAULT_FILL_TYPE;

    }

    @Override
    protected void onTextChanged(CharSequence text, int start, int lengthBefore, int lengthAfter) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter);
        /** 纯色小红点模式下若有文本需要从无变为有, 要归位view的大小.*/
        /** Если в режиме сплошной красной точки есть текст, вам нужно изменить размер представления. */
        String strText = text==null?"":text.toString().trim();
        if(isHighLightMode && !"".equals(strText)){
            ViewGroup.LayoutParams lp = getLayoutParams();
            lp.height = ViewGroup.LayoutParams.WRAP_CONTENT;
            lp.width = ViewGroup.LayoutParams.WRAP_CONTENT;
            setLayoutParams(lp);
            isHighLightMode = false;
        }
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        refreshBackgroundDrawable(w, h);
    }

    private void refreshBackgroundDrawable(int targetWidth, int targetHeight) {
        if (targetWidth <= 0 || targetHeight <= 0) {
            return;
        }
        CharSequence text = getText();
        if (text == null) {
            return;
        }
        if (text.length() == 1) {
            /**第一种背景是一个正圆形, 当文本为个位数字时 */
            /** Первый фон представляет собой идеальный круг, когда текст представляет собой однозначное число */
            int max = Math.max(targetWidth, targetHeight);
            ShapeDrawable circle;
            final int diameter = max - (2 * mShadowRadius);
            OvalShape oval = new OvalShadow(mShadowRadius, diameter);
            circle = new ShapeDrawable(oval);
            // AA ViewCompat.setLayerType(this, ViewCompat.LAYER_TYPE_SOFTWARE, circle.getPaint());
            ViewCompat.setLayerType(this, View.LAYER_TYPE_SOFTWARE, circle.getPaint());
            circle.getPaint().setShadowLayer(mShadowRadius, shadowXOffset, shadowYOffset, KEY_SHADOW_COLOR);
            circle.getPaint().setColor(backgroundColor);
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.JELLY_BEAN) {
                setBackgroundDrawable(circle);
            } else {
                setBackground(circle);
            }
        } else if (text.length() > 1) {
            /**第二种背景是上下两边为直线的椭圆, 当文本长度大于1时 */
            /** Вторым фоном является эллипс с прямыми линиями сверху и снизу. Когда длина текста больше 1, */
            SemiCircleRectDrawable sr = new SemiCircleRectDrawable();
            //ViewCompat.setLayerType(this, ViewCompat.LAYER_TYPE_SOFTWARE, sr.getPaint());
            ViewCompat.setLayerType(this, View.LAYER_TYPE_SOFTWARE, sr.getPaint());
            sr.getPaint().setShadowLayer(mShadowRadius, shadowXOffset, shadowYOffset, KEY_SHADOW_COLOR);
            sr.getPaint().setColor(backgroundColor);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                setBackground(sr);
            } else {
                setBackgroundDrawable(sr);
            }
        } else {
            /** 第三种情况就是text="", 即文本长度为0, 因为无任何文本, 则对当前的TextView背景不做任何更新,
             * 但是有时候我们需要一个无字的纯色小圆形,用来表达强调.这种情况因为要重新设置View的大小, 所以不在这里表现, 请使用另外一个方法setHighLightMode()来完成.
             */
            /** Третий случай - это text = "", то есть длина текста равна 0.
             *   Поскольку текст отсутствует, обновления для текущего фона TextView не производятся.
             *   Но иногда нам нужен сплошной маленький кружок без слов, чтобы выразить акцент.
             *   Это потому, что размер представления сбрасывается, поэтому он здесь не представлен.
             *   Пожалуйста, используйте другой метод setHighLightMode () для завершения.
             */

        }

    }

    public void setBadgeCount(String count){
        setBadgeCount(count, false);
    }
    public void setBadgeCount(String count, boolean goneWhenZero) {
        int temp = -1;
        try {
            temp = Integer.parseInt(count);
        } catch (Exception e) {
            Log.d("Warning", e.getStackTrace()[0].getMethodName() + " : " + e.toString());
        }
        if (temp != -1) {
            setBadgeCount(temp, goneWhenZero);
        }
    }

    public void setBadgeCount(int count){
        setBadgeCount(count, true);
    }
    public void setBadgeCount(int count, boolean goneWhenZero){
        if(count >0 && count <= 99){
            setText(String.valueOf(count));
            setVisibility(View.VISIBLE);
        }else if(count >99){
            setText("99+");
            setVisibility(View.VISIBLE);
        }else if(count <= 0){
            setText("0");
            if(goneWhenZero){
                setVisibility(View.GONE);
            }else{
                setVisibility(View.VISIBLE);
            }
        }
    }

    /**
     * 明确的展现一个无任何文本的红色圆点,
     * 主要是通过设置文本setText("")触发onTextChanged(), 再连锁触发onSizeChanged()最后更新了背景.
     * Четко показать красную точку без текста,
     * Главным образом, вызывая onTextChanged (), устанавливая текст setText (""),
     * затем связывая цепочку onSizeChanged (), чтобы, наконец, обновить фон.
     */
    public void setHighLightMode(){
        setHighLightMode(false);
    }

    public void clearHighLightMode(){
        isHighLightMode = false;
        setBadgeCount(0);
    }

    /**
     *
     * @param isDisplayInToolbarMenu
     */
    public void setHighLightMode(boolean isDisplayInToolbarMenu){
        isHighLightMode = true;
        ViewGroup.LayoutParams params = getLayoutParams();
        params.width = dp2px(getContext(), 8);
        params.height = params.width;
        if(isDisplayInToolbarMenu && params instanceof FrameLayout.LayoutParams){
            ((FrameLayout.LayoutParams)params).topMargin=dp2px(getContext(), 8);
            ((FrameLayout.LayoutParams)params).rightMargin=dp2px(getContext(), 8);
        }
        setLayoutParams(params);
        ShapeDrawable drawable = new ShapeDrawable(new OvalShape());
        // AA ViewCompat.setLayerType(this, ViewCompat.LAYER_TYPE_SOFTWARE, drawable.getPaint());
        ViewCompat.setLayerType(this, View.LAYER_TYPE_SOFTWARE, drawable.getPaint());
        drawable.getPaint().setColor(backgroundColor);
        drawable.getPaint().setAntiAlias(true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            setBackground(drawable);
        } else {
            setBackgroundDrawable(drawable);
        }
        setText("");
        setVisibility(View.VISIBLE);
    }

    public void setBackgroundColor(int color){
        backgroundColor = color;
        refreshBackgroundDrawable(getWidth(), getHeight());
    }

    private class OvalShadow extends OvalShape {
        private final Paint mShadowPaint;
        private final int mCircleDiameter;

        public OvalShadow(int shadowRadius, int circleDiameter) {
            super();
            mShadowPaint = new Paint();
            mShadowRadius = shadowRadius;
            mCircleDiameter = circleDiameter;
            RadialGradient mRadialGradient = new RadialGradient(mCircleDiameter / 2, mCircleDiameter / 2,
                    mShadowRadius, new int[]{
                    FILL_SHADOW_COLOR, Color.TRANSPARENT
            }, null, Shader.TileMode.CLAMP);
            mShadowPaint.setShader(mRadialGradient);
        }

        @Override
        public void draw(Canvas canvas, Paint paint) {
            final int viewWidth = MaterialBadgeTextView.this.getWidth();
            final int viewHeight = MaterialBadgeTextView.this.getHeight();
            canvas.drawCircle(viewWidth / 2, viewHeight / 2, (mCircleDiameter / 2 + mShadowRadius), mShadowPaint);
            canvas.drawCircle(viewWidth / 2, viewHeight / 2, (mCircleDiameter / 2), paint);
        }
    }

    class SemiCircleRectDrawable extends Drawable {
        private final Paint mPaint;
        private RectF rectF;

        public Paint getPaint() {
            return mPaint;
        }

        public SemiCircleRectDrawable() {
            mPaint = new Paint();
            mPaint.setAntiAlias(true);
        }

        @Override
        public void setBounds(int left, int top, int right, int bottom) {
            super.setBounds(left, top, right, bottom);
            if (rectF == null) {
                rectF = new RectF(left + diffWH, top + mShadowRadius+4, right - diffWH, bottom - mShadowRadius-4);
            } else {
                rectF.set(left + diffWH, top + mShadowRadius+4, right - diffWH, bottom - mShadowRadius-4);
            }
        }

        @Override
        public void draw(Canvas canvas) {
            float R = (float)(rectF.bottom * 0.4);
            if (rectF.right < rectF.bottom) {
                R = (float)(rectF.right * 0.4);
            }
            canvas.drawRoundRect(rectF, R, R, mPaint);
        }

        @Override
        public void setAlpha(int alpha) {
            mPaint.setAlpha(alpha);
        }

        @Override
        public void setColorFilter(ColorFilter colorFilter) {
            mPaint.setColorFilter(colorFilter);
        }

        @Override
        public int getOpacity() {
            return PixelFormat.TRANSPARENT;
        }
    }

    public static int dp2px(Context context, float dpValue) {
        try {
            final float scale = context.getResources().getDisplayMetrics().density;
            return (int) (dpValue * scale + 0.5f);
        } catch (Exception e) {
            return (int) (dpValue + 0.5f);
        }
    }

}

File: app4web_k1/app/src/main/java/site/app4web/app4web/Lib/MaterialBadgeTextView.kt
// A mildly-modified version of the MaterialBadgeTextView library from:
// https://github.com/matrixxun/MaterialBadgeTextView
package site.app4web.app4web.Lib

import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.OvalShape
import android.os.Build
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.view.ViewCompat
import site.app4web.app4web.Helper.JasonHelper

//import android.support.v7.widget.AppCompatTextView;
/**
 * Created by matrixxun on 2016/8/30.
 */
class MaterialBadgeTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr) {
    private var backgroundColor = 0
    private var borderColor = 0
    private var borderWidth = 0f
    private var borderAlpha = 0f
    private var ctType = 0
    private var density = 0f
    private var mShadowRadius = 0
    private var shadowYOffset = 0
    private var shadowXOffset = 0
    private var basePadding = 0
    private var diffWH = 0
    private var isHighLightMode = false
    private fun init(context: Context, attrs: AttributeSet?) {
        gravity = Gravity.CENTER
        density = getContext().resources.displayMetrics.density
        mShadowRadius = (density * SHADOW_RADIUS).toInt()
        shadowYOffset = (density * Y_OFFSET).toInt()
        shadowXOffset = (density * X_OFFSET).toInt()
        basePadding = mShadowRadius * 2
        val textHeight = textSize
        val textWidth = textHeight / 4
        diffWH = (Math.abs(textHeight - textWidth) / 2).toInt()
        val horizontalPadding = basePadding + diffWH
        setPadding(horizontalPadding, basePadding, horizontalPadding, basePadding)
        backgroundColor = JasonHelper.parse_color("#ff0000")
        borderColor = Color.TRANSPARENT
        borderWidth = 0f
        borderAlpha = 1f
        ctType = DEFAULT_FILL_TYPE
    }

    override fun onTextChanged(
        text: CharSequence,
        start: Int,
        lengthBefore: Int,
        lengthAfter: Int
    ) {
        super.onTextChanged(text, start, lengthBefore, lengthAfter)
        /** 纯色小红点模式下若有文本需要从无变为有, 要归位view的大小. */
        /** Если в режиме сплошной красной точки есть текст, вам нужно изменить размер представления.  */
        val strText = text.toString().trim { it <= ' ' }
        if (isHighLightMode && "" != strText) {
            val lp = layoutParams
            lp.height = ViewGroup.LayoutParams.WRAP_CONTENT
            lp.width = ViewGroup.LayoutParams.WRAP_CONTENT
            layoutParams = lp
            isHighLightMode = false
        }
    }

    override fun onSizeChanged(
        w: Int,
        h: Int,
        oldw: Int,
        oldh: Int
    ) {
        super.onSizeChanged(w, h, oldw, oldh)
        refreshBackgroundDrawable(w, h)
    }

    private fun refreshBackgroundDrawable(targetWidth: Int, targetHeight: Int) {
        if (targetWidth <= 0 || targetHeight <= 0) {
            return
        }
        val text = text ?: return
        if (text.length == 1) {
            /**第一种背景是一个正圆形, 当文本为个位数字时  */
            /** Первый фон представляет собой идеальный круг, когда текст представляет собой однозначное число  */
            val max = Math.max(targetWidth, targetHeight)
            val circle: ShapeDrawable
            val diameter = max - 2 * mShadowRadius
            val oval: OvalShape =
                OvalShadow(mShadowRadius, diameter)
            circle = ShapeDrawable(oval)
            // AA ViewCompat.setLayerType(this, ViewCompat.LAYER_TYPE_SOFTWARE, circle.getPaint());
            ViewCompat.setLayerType(this, View.LAYER_TYPE_SOFTWARE, circle.paint)
            circle.paint.setShadowLayer(
                mShadowRadius.toFloat(),
                shadowXOffset.toFloat(),
                shadowYOffset.toFloat(),
                KEY_SHADOW_COLOR
            )
            circle.paint.color = backgroundColor
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.JELLY_BEAN) {
                setBackgroundDrawable(circle)
            } else {
                background = circle
            }
        } else if (text.length > 1) {
            /**第二种背景是上下两边为直线的椭圆, 当文本长度大于1时  */
            /** Вторым фоном является эллипс с прямыми линиями сверху и снизу. Когда длина текста больше 1,  */
            val sr = SemiCircleRectDrawable()
            //ViewCompat.setLayerType(this, ViewCompat.LAYER_TYPE_SOFTWARE, sr.getPaint());
            ViewCompat.setLayerType(this, View.LAYER_TYPE_SOFTWARE, sr.paint)
            sr.paint.setShadowLayer(
                mShadowRadius.toFloat(),
                shadowXOffset.toFloat(),
                shadowYOffset.toFloat(),
                KEY_SHADOW_COLOR
            )
            sr.paint.color = backgroundColor
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                background = sr
            } else {
                setBackgroundDrawable(sr)
            }
        } else {
            /** 第三种情况就是text="", 即文本长度为0, 因为无任何文本, 则对当前的TextView背景不做任何更新,
             * 但是有时候我们需要一个无字的纯色小圆形,用来表达强调.这种情况因为要重新设置View的大小, 所以不在这里表现, 请使用另外一个方法setHighLightMode()来完成.
             */
            /** Третий случай - это text = "", то есть длина текста равна 0.
             * Поскольку текст отсутствует, обновления для текущего фона TextView не производятся.
             * Но иногда нам нужен сплошной маленький кружок без слов, чтобы выразить акцент.
             * Это потому, что размер представления сбрасывается, поэтому он здесь не представлен.
             * Пожалуйста, используйте другой метод setHighLightMode () для завершения.
             */
        }
    }

    fun setBadgeCount(count: String) {
        setBadgeCount(count, false)
    }

    fun setBadgeCount(count: String, goneWhenZero: Boolean) {
        var temp = -1
        try {
            temp = count.toInt()
        } catch (e: Exception) {
            Log.d(
                "Warning",
                e.stackTrace[0].methodName + " : " + e.toString()
            )
        }
        if (temp != -1) {
            setBadgeCount(temp, goneWhenZero)
        }
    }

    fun setBadgeCount(count: Int) {
        setBadgeCount(count, true)
    }

    fun setBadgeCount(count: Int, goneWhenZero: Boolean) {
        if (count > 0 && count <= 99) {
            text = count.toString()
            visibility = View.VISIBLE
        } else if (count > 99) {
            text = "99+"
            visibility = View.VISIBLE
        } else if (count <= 0) {
            text = "0"
            visibility = if (goneWhenZero) {
                View.GONE
            } else {
                View.VISIBLE
            }
        }
    }

    /**
     * 明确的展现一个无任何文本的红色圆点,
     * 主要是通过设置文本setText("")触发onTextChanged(), 再连锁触发onSizeChanged()最后更新了背景.
     * Четко показать красную точку без текста,
     * Главным образом, вызывая onTextChanged (), устанавливая текст setText (""),
     * затем связывая цепочку onSizeChanged (), чтобы, наконец, обновить фон.
     */
    fun setHighLightMode() {
        setHighLightMode(false)
    }

    fun clearHighLightMode() {
        isHighLightMode = false
        setBadgeCount(0)
    }

    /**
     *
     * @param isDisplayInToolbarMenu
     */
    fun setHighLightMode(isDisplayInToolbarMenu: Boolean) {
        isHighLightMode = true
        val params = layoutParams
        params.width = dp2px(context, 8f)
        params.height = params.width
        if (isDisplayInToolbarMenu && params is FrameLayout.LayoutParams) {
            params.topMargin =
                dp2px(context, 8f)
            params.rightMargin =
                dp2px(context, 8f)
        }
        layoutParams = params
        val drawable = ShapeDrawable(OvalShape())
        // AA ViewCompat.setLayerType(this, ViewCompat.LAYER_TYPE_SOFTWARE, drawable.getPaint());
        ViewCompat.setLayerType(this, View.LAYER_TYPE_SOFTWARE, drawable.paint)
        drawable.paint.color = backgroundColor
        drawable.paint.isAntiAlias = true
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            background = drawable
        } else {
            setBackgroundDrawable(drawable)
        }
        text = ""
        visibility = View.VISIBLE
    }

    override fun setBackgroundColor(color: Int) {
        backgroundColor = color
        refreshBackgroundDrawable(width, height)
    }

    private inner class OvalShadow(
        shadowRadius: Int,
        circleDiameter: Int
    ) : OvalShape() {
        private val mRadialGradient: RadialGradient
        private val mShadowPaint: Paint
        private val mCircleDiameter: Int
        override fun draw(canvas: Canvas, paint: Paint) {
            val viewWidth = this@MaterialBadgeTextView.width
            val viewHeight = this@MaterialBadgeTextView.height
            canvas.drawCircle(
                viewWidth / 2.toFloat(),
                viewHeight / 2.toFloat(),
                (mCircleDiameter / 2 + mShadowRadius).toFloat(),
                mShadowPaint
            )
            canvas.drawCircle(
                viewWidth / 2.toFloat(),
                viewHeight / 2.toFloat(),
                (mCircleDiameter / 2).toFloat(),
                paint
            )
        }

        init {
            mShadowPaint = Paint()
            mShadowRadius = shadowRadius
            mCircleDiameter = circleDiameter
            mRadialGradient = RadialGradient(
                (mCircleDiameter / 2).toFloat(), (mCircleDiameter / 2).toFloat(),
                mShadowRadius.toFloat(), intArrayOf(
                    FILL_SHADOW_COLOR, Color.TRANSPARENT
                ), null, Shader.TileMode.CLAMP
            )
            mShadowPaint.shader = mRadialGradient
        }
    }

    internal inner class SemiCircleRectDrawable : Drawable() {
        val paint: Paint
        private var rectF: RectF? = null

        override fun setBounds(
            left: Int,
            top: Int,
            right: Int,
            bottom: Int
        ) {
            super.setBounds(left, top, right, bottom)
            if (rectF == null) {
                rectF = RectF(
                    (left + diffWH).toFloat(),
                    (top + mShadowRadius + 4).toFloat(),
                    (right - diffWH).toFloat(),
                    (bottom - mShadowRadius - 4).toFloat()
                )
            } else {
                rectF!![left + diffWH.toFloat(), top + mShadowRadius + 4.toFloat(), right - diffWH.toFloat()] =
                    bottom - mShadowRadius - 4.toFloat()
            }
        }

        override fun draw(canvas: Canvas) {
            var R = (rectF!!.bottom * 0.4).toFloat()
            if (rectF!!.right < rectF!!.bottom) {
                R = (rectF!!.right * 0.4).toFloat()
            }
            canvas.drawRoundRect(rectF!!, R, R, paint)
        }

        override fun setAlpha(alpha: Int) {
            paint.alpha = alpha
        }

        override fun setColorFilter(colorFilter: ColorFilter?) {
            paint.colorFilter = colorFilter
        }

        override fun getOpacity(): Int {
            return PixelFormat.TRANSPARENT
        }

        init {
            paint = Paint()
            paint.isAntiAlias = true
        }
    }

    companion object {
        private const val DEFAULT_FILL_TYPE = 0
        private const val SHADOW_RADIUS = 3.5f
        private const val FILL_SHADOW_COLOR = 0x55000000
        private const val KEY_SHADOW_COLOR = 0x55000000
        private const val X_OFFSET = 0f
        private const val Y_OFFSET = 1.75f
        fun dp2px(context: Context, dpValue: Float): Int {
            return try {
                val scale = context.resources.displayMetrics.density
                (dpValue * scale + 0.5f).toInt()
            } catch (e: Exception) {
                (dpValue + 0.5f).toInt()
            }
        }
    }

    init {
        init(context, attrs)
    }
}